File: android/beaconsdk/src/main/java/com/xenber/beaconsdk/notification/NotificationFactory.kt
package com.xenber.beaconsdk.notification

import android.app.NotificationManager
import android.app.NotificationChannel
import android.os.Build
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.net.Uri
import androidx.core.app.NotificationCompat
// import com.xenber.beaconsdk.R
import com.xenber.beaconsdk.util.Logger
import org.json.JSONObject

object NotificationFactory {

    private const val ALERT_CHANNEL = "beacon_detection_alert"
    private const val EXTERNAL_TRIGGER_NOTI_ID = 1001

    // fun triggerExternalApp(
    //     context: Context,
    //     mac: String,
    //     targetApp: String,
    //     params: JSONObject
    // ) {
    //     try {
    //         val uriBuilder = Uri.parse(targetApp).buildUpon()
    //         params.keys().forEach {
    //             uriBuilder.appendQueryParameter(it, params.get(it).toString())
    //         }

    //         val intent = Intent(Intent.ACTION_VIEW, uriBuilder.build()).apply {
    //             flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
    //         }

    //         val pi = PendingIntent.getActivity(
    //             context,
    //             mac.hashCode(),
    //             intent,
    //             PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
    //         )

    //         val notification = NotificationCompat.Builder(context, ALERT_CHANNEL)
    //             .setSmallIcon(android.R.drawable.ic_dialog_info)
    //             .setContentTitle(" Nearby: ${params.optString("loc")}")
    //             .setContentText("Tap to check-in for ${params.optString("shift")} shift")
    //             .setPriority(NotificationCompat.PRIORITY_MAX)
    //             .setContentIntent(pi)
    //             .setAutoCancel(true)
    //             .build()

    //         val nm = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
    //         nm.notify(EXTERNAL_TRIGGER_NOTI_ID, notification)

    //     } catch (e: Exception) {
    //         Logger.e("Trigger failed: ${e.message}")
    //     }
    // }

    fun showInternalNotification(
        context: Context,
        mac: String,
        params: JSONObject
    ){
        try {
            
            val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                val channel = NotificationChannel(
                    ALERT_CHANNEL,
                    "Beacon Alerts",
                    NotificationManager.IMPORTANCE_HIGH
                ).apply {
                description = "Beacons Detected "
            }
                notificationManager.createNotificationChannel(channel)
            }

            val launchIntent =
            context.packageManager.getLaunchIntentForPackage(context.packageName)
                ?: run {
                    Logger.e("SDK Error: Could not resolve host app Launcher Activity.")
                    return
                }

            launchIntent.apply {
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP)
                putExtra("action", "beacon_detected")
                putExtra("mac_address", mac)
                putExtra("location_name", params.optString("loc", "Unknown Location"))
                putExtra("rssi", params.optInt("rssi", 0))
                putExtra("timestamp", params.optLong("timestamp", 0L))
            }

            val pi = PendingIntent.getActivity(
                context,
                mac.hashCode(),
                launchIntent,
                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
            )

            val notification = NotificationCompat.Builder(context, ALERT_CHANNEL)
                .setSmallIcon(android.R.drawable.ic_dialog_info)
                .setContentTitle(" Nearby: ${params.optString("loc", "Unknown Location")}")
                .setContentText("Tap to check-in for ${params.optString("shift", "your shift")} shift")
                .setPriority(NotificationCompat.PRIORITY_MAX)
                .setContentIntent(pi)
                .setAutoCancel(true)
                .build()

            notificationManager.notify(mac.hashCode(), notification)

        } catch (e: Exception) {
            Logger.e("Notification failed: ${e.message}")
        }
    }
}

File: ios/BeaconSdk/Notification/NotificationFactory.swift
import Foundation
import UserNotifications
import UIKit

class NotificationFactory: NSObject {
  static let shared = NotificationFactory()
  private override init() {
    super.init()
  }

  func requestPermission() {
    let center = UNUserNotificationCenter.current()
    center.requestAuthorization(options:[.alert, .sound, .badge]) {granted, error in
      if let error = error {
        Logger.e("Notification permission error", error)
      } else if granted {
        Logger.i("Notification permission granted")
      } 
    }
  }

  func showInternalNotification(mac: String, params:[String: Any]) {

    let locationName = params["loc"] as? String ?? "Unknown Location"
    let shiftName = params["shift"] as? String ?? "your shift"
    let rssi = params["rssi"] as? Int ?? 0
    let timestamp = params["timestamp"] as? Int64 ?? 0

    let content = UNMutableNotificationContent()
    content.title = " Nearby: \(locationName)"
    content.body = "Tap to check-in for \(shiftName) shift"
    content.sound = .default

    content.userInfo = [
      "action": "beacon_detected",
      "mac_address": mac,
      "location_name": locationName,
      "rssi": rssi,
      "timestamp": timestamp
    ]

    let trigger = UNTimeIntervalNotificationTrigger(timeInterval: 1, repeats: false)

    let identifier = String(mac.hashValue)
    let request = UNNotificationRequest(identifier: identifier, content: content, trigger: trigger)

    UNUserNotificationCenter.current().add(request) { error in
      if let error = error {
        Logger.e("Notification failed", error)
      } else {
        Logger.d("Notification posted for \(locationName)")
      }
    }
  }
}